### ISSUE TITLE
XWiki release blocked by some weird revapi violations

### ISSUE THREAD

**Issue by vmassol** [revapi/revapi#114]
Hi (Lukas),

We have a problem and been stuck with it for the past several days (and this is blocking our XWiki release ATM).

In short when building `xwiki-platform/xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-script`, we get:

```
[ERROR] Failed to execute goal org.revapi:revapi-maven-plugin:0.10.0:check (revapi-check) on project xwiki-platform-extension-script: The following API problems caused the build to fail:
[ERROR] java.method.finalMethodAddedToNonFinalClass: method java.lang.Class<?> java.lang.Object::getClass() @ org.xwiki.component.namespace.Namespace: A final method has been added to an inheritable class. [<system classpath>]
[ERROR] java.method.finalMethodAddedToNonFinalClass: method void java.lang.Object::notify() @ org.xwiki.component.namespace.Namespace: A final method has been added to an inheritable class. [<system classpath>]
[ERROR] java.method.finalMethodAddedToNonFinalClass: method void java.lang.Object::notifyAll() @ org.xwiki.component.namespace.Namespace: A final method has been added to an inheritable class. [<system classpath>]
[ERROR] java.method.finalMethodAddedToNonFinalClass: method void java.lang.Object::wait() throws java.lang.InterruptedException @ org.xwiki.component.namespace.Namespace: A final method has been added to an inheritable class. [<system classpath>]
[ERROR] java.method.finalMethodAddedToNonFinalClass: method void java.lang.Object::wait(long) throws java.lang.InterruptedException @ org.xwiki.component.namespace.Namespace: A final method has been added to an inheritable class. [<system classpath>]
[ERROR] java.method.finalMethodAddedToNonFinalClass: method void java.lang.Object::wait(long, int) throws java.lang.InterruptedException @ org.xwiki.component.namespace.Namespace: A final method has been added to an inheritable class. [<system classpath>]
[ERROR] 
[ERROR] If you're using the semver-ignore extension, update your module's version to one compatible with the current changes (e.g. mvn package revapi:update-versions). If you want to explicitly ignore this change and provide a justification for it, add the following JSON snippet to your Revapi configuration under "revapi.ignore" path:
[ERROR] {
[ERROR]   "code": "java.method.finalMethodAddedToNonFinalClass",
[ERROR]   "new": "method java.lang.Class<?> java.lang.Object::getClass() @ org.xwiki.component.namespace.Namespace",
[ERROR]   "package": "org.xwiki.component.namespace",
[ERROR]   "classSimpleName": "Namespace",
[ERROR]   "methodName": "getClass",
[ERROR]   "elementKind": "method",
[ERROR]   "justification": <<<<< ADD YOUR EXPLANATION FOR THE NECESSITY OF THIS CHANGE >>>>>
[ERROR] },
[ERROR] {
[ERROR]   "code": "java.method.finalMethodAddedToNonFinalClass",
[ERROR]   "new": "method void java.lang.Object::notify() @ org.xwiki.component.namespace.Namespace",
[ERROR]   "package": "org.xwiki.component.namespace",
[ERROR]   "classSimpleName": "Namespace",
[ERROR]   "methodName": "notify",
[ERROR]   "elementKind": "method",
[ERROR]   "justification": <<<<< ADD YOUR EXPLANATION FOR THE NECESSITY OF THIS CHANGE >>>>>
[ERROR] },
[ERROR] {
[ERROR]   "code": "java.method.finalMethodAddedToNonFinalClass",
[ERROR]   "new": "method void java.lang.Object::notifyAll() @ org.xwiki.component.namespace.Namespace",
[ERROR]   "package": "org.xwiki.component.namespace",
[ERROR]   "classSimpleName": "Namespace",
[ERROR]   "methodName": "notifyAll",
[ERROR]   "elementKind": "method",
[ERROR]   "justification": <<<<< ADD YOUR EXPLANATION FOR THE NECESSITY OF THIS CHANGE >>>>>
[ERROR] },
[ERROR] {
[ERROR]   "code": "java.method.finalMethodAddedToNonFinalClass",
[ERROR]   "new": "method void java.lang.Object::wait() throws java.lang.InterruptedException @ org.xwiki.component.namespace.Namespace",
[ERROR]   "package": "org.xwiki.component.namespace",
[ERROR]   "classSimpleName": "Namespace",
[ERROR]   "methodName": "wait",
[ERROR]   "elementKind": "method",
[ERROR]   "justification": <<<<< ADD YOUR EXPLANATION FOR THE NECESSITY OF THIS CHANGE >>>>>
[ERROR] },
[ERROR] {
[ERROR]   "code": "java.method.finalMethodAddedToNonFinalClass",
[ERROR]   "new": "method void java.lang.Object::wait(long) throws java.lang.InterruptedException @ org.xwiki.component.namespace.Namespace",
[ERROR]   "package": "org.xwiki.component.namespace",
[ERROR]   "classSimpleName": "Namespace",
[ERROR]   "methodName": "wait",
[ERROR]   "elementKind": "method",
[ERROR]   "justification": <<<<< ADD YOUR EXPLANATION FOR THE NECESSITY OF THIS CHANGE >>>>>
[ERROR] },
[ERROR] {
[ERROR]   "code": "java.method.finalMethodAddedToNonFinalClass",
[ERROR]   "new": "method void java.lang.Object::wait(long, int) throws java.lang.InterruptedException @ org.xwiki.component.namespace.Namespace",
[ERROR]   "package": "org.xwiki.component.namespace",
[ERROR]   "classSimpleName": "Namespace",
[ERROR]   "methodName": "wait",
[ERROR]   "elementKind": "method",
[ERROR]   "justification": <<<<< ADD YOUR EXPLANATION FOR THE NECESSITY OF THIS CHANGE >>>>>
[ERROR] }
```

Which is strange on 2 accounts:

1) We explicitly say in platform not to take care of problems from commons or rendering with:

```
          <analysisConfiguration><![CDATA[
            {
              "revapi": {
                "filter": {
                  // We're ignoring all API incompatibilities from XWiki dependencies (i.e. commons and rendering)
                  // since they're already covered by their own builds and we release all of them together.
                  "archives": {
                    "exclude": ["org\\.xwiki\\.commons:.*:.*", "org\\.xwiki\\.rendering:.*:.*"]
                  }
                },
                "ignore" : [
...
```

2) Even for commons, these errors don't make sense. For example:

```
[ERROR] {
[ERROR]   "code": "java.method.finalMethodAddedToNonFinalClass",
[ERROR]   "new": "method java.lang.Class<?> java.lang.Object::getClass() @ org.xwiki.component.namespace.Namespace",
[ERROR]   "package": "org.xwiki.component.namespace",
[ERROR]   "classSimpleName": "Namespace",
[ERROR]   "methodName": "getClass",
[ERROR]   "elementKind": "method",
[ERROR]   "justification": <<<<< ADD YOUR EXPLANATION FOR THE NECESSITY OF THIS CHANGE >>>>>
[ERROR] },
```

You can check the Namespace class at https://github.com/xwiki/xwiki-commons/blob/6b71a9481702a502ce31892abd6df1025544d0cc/xwiki-commons-core/xwiki-commons-component/xwiki-commons-component-api/src/main/java/org/xwiki/component/namespace/Namespace.java#L31

Thanks for your help Lukas!

**Comment by metlos** [comment]
Interesting... I would have understood it if those violations were accompanied by the change of the class type (i.e. from interface to an actual class) but that doesn't seem to be the case here...

**Comment by vmassol** [comment]
We have tons of those, in various modules (just found a lot more). In order to be able to release, I'm trying to define a global ignore for them.

Found a global regex to exclude them all:

```
                  {
                    "regex" : true,
                    "code": ".*",
                    "new": ".*java\..* @.*",
                    "justification": "Bug in revapi"
                  }
````

**Comment by vmassol** [comment]
Something strange, whenever I run `mvn revapi:check` maven is redownloading stuff and I'd have expected them to be already downloaded since I'm executing the same command on the same module several times. For example:

```
[INFO] --- revapi-maven-plugin:0.10.0:check (default-cli) @ xwiki-platform-image-plugin ---
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public/org/xwiki/platform/xwiki-platform-image-plugin/9.11.1/xwiki-platform-image-plugin-9.11.1.jar
Downloading from : https://repo.maven.apache.org/maven2/org/xwiki/platform/xwiki-platform-image-plugin/9.11.1/xwiki-platform-image-plugin-9.11.1.jar
[WARNING] Failed to resolve old artifacts: Could not find artifact org.xwiki.platform:xwiki-platform-image-plugin:jar:9.11.1 in xwiki-releases (http://nexus.xwiki.org/nexus/content/groups/public). The API analysis will proceed comparing the new archives against an empty archive.
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/platform/xwiki-platform-image-plugin/10.0-SNAPSHOT/maven-metadata.xml
Downloaded from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/platform/xwiki-platform-image-plugin/10.0-SNAPSHOT/maven-metadata.xml (795 B at 11 kB/s)
Downloading from : https://repo.maven.apache.org/maven2/io/sf/jclf/jclf/maven-metadata.xml
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/io/sf/jclf/jclf/maven-metadata.xml
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public/io/sf/jclf/jclf/maven-metadata.xml
Downloaded from : http://nexus.xwiki.org/nexus/content/groups/public/io/sf/jclf/jclf/maven-metadata.xml (321 B at 7.1 kB/s)
[WARNING] The POM for com.sun.speech.freetts:freetts:jar:1.2.1 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-component-default/10.0-SNAPSHOT/maven-metadata.xml
Downloaded from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-component-default/10.0-SNAPSHOT/maven-metadata.xml (798 B at 22 kB/s)
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-component-observation/10.0-SNAPSHOT/maven-metadata.xml
Downloaded from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-component-observation/10.0-SNAPSHOT/maven-metadata.xml (802 B at 22 kB/s)
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-component-api/10.0-SNAPSHOT/maven-metadata.xml
Downloaded from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-component-api/10.0-SNAPSHOT/maven-metadata.xml (794 B at 11 kB/s)
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-stability/10.0-SNAPSHOT/maven-metadata.xml
Downloaded from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-stability/10.0-SNAPSHOT/maven-metadata.xml (790 B at 9.5 kB/s)
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-text/10.0-SNAPSHOT/maven-metadata.xml
Downloaded from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-text/10.0-SNAPSHOT/maven-metadata.xml (785 B at 11 kB/s)
Downloading from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-environment-api/10.0-SNAPSHOT/maven-metadata.xml
Downloaded from : http://nexus.xwiki.org/nexus/content/groups/public-snapshots/org/xwiki/commons/xwiki-commons-environment-api/10.0-SNAPSHOT/maven-metadata.xml (796 B at 23 kB/s)
...
```

**Comment by vmassol** [comment]
One other thing that is weird is that revapi is still doing stuff on modules that don't have previous versions such as `xwiki-platform-image-plugin`. It should just stop.

**Comment by vmassol** [comment]
Just tested with 0.9.5 and I don't get those errors so it seems to be a regression of 0.10.0.

**Comment by metlos** [comment]
This seems to be fallout from #109. I changed the logic so that SOME of the checks should run even if there is no prior version. But by the looks of it, it doesn't work as expected.

**Comment by metlos** [comment]
> Note that we upgraded to 0.10.0 to try to fix another problem which in the end was because revapi only works with Maven 3.5 now. 

Could you tell me more about that? I'm not sure I raised that requirement intentionally.

> So we'll probably go back to 0.9.5 now till this is issue is fixed.

:+1: Glad that solves it for you for the time being.

**Comment by vmassol** [comment]
> Could you tell me more about that? I'm not sure I raised that requirement intentionally.

We had xwiki-platform-distribution failing with some revapi errors (even though we used the `<skip>true</skip>` parameter in those modules BTW) that were caused by the fact on our CI we were using Maven 3.3.9.

So this forced us to now make XWiki buildable only with Maven 3.5.0+ ( see http://dev.xwiki.org/xwiki/bin/view/Community/Building/?viewer=changes&rev1=103.1&rev2=104.1 ).

I don't have the error at hand but you should be able to reproduce by running `mvn revapi:check` on https://github.com/xwikisas/xwiki-platform/tree/master/xwiki-platform-distribution using Maven 3.3.9.

Thanks!

**Comment by tmortagne** [comment]
You can find the result of xwiki-platform-distribution build with 3.3.9 on http://ci.xwiki.org/blue/organizations/jenkins/XWiki%2Fxwiki-platform/detail/master/362/pipeline/.

But anyway IMO the plugin should have as dependency the minimum version of Maven it wants to support. There is no reason to expect a plugin with a dependency on Maven 3.5 API to fully work with a lower version of Maven.

3.3.9 is a good number since that's the version you got in current Ubuntu LTS for example.

**Comment by metlos** [comment]
Well, that's the thing - the minimum supported maven version should be 3.2.5... The plugin should not depend on anything maven 3.5 specific as far as I know...

**Comment by tmortagne** [comment]
Actually I looked too quickly at <maven-plugin-plugin.version> which is not synchronized with Maven version sorry. So now sure where this error come from exactly then but it really suggest that 3.5 is involved somewhere, for example org.apache.maven.repository.internal.DefaultVersionRangeResolver.logger is org.eclipse.aether.spi.log.Logger in 3.3.9 but for some reason it's trying to inject org.eclipse.aether.internal.impl.slf4j.Slf4jLoggerFactory which is one of the changes in 3.5 (the move to slf4j).

**Comment by tmortagne** [comment]
Hmm after looking at the plugin dependencies I'm wondering if it would not be related to aether version. The plugin "forces" aether-api 0.9.1 while maven-aether-provider 3.2.5 expect 1.0.0.

**Comment by metlos** [comment]
I'm closing this issue since the fix blocking the use of `revapi-maven-plugin-0.10.0` is fixed in `revapi-java-0.15.1`.

I've opened #116 to track the Maven version problems.
